### ISSUE TITLE
Type narrowing infers `object` when `isinstance` is used in a ternary operation

### ISSUE THREAD

**Issue by HosseyNJF** [astral-sh/ty#3522]
### Summary

In short, this line of code doesn't work as expected:

```python3
# Neither of these work
single: SomeClass = seq_or_single[0] if isinstance(seq_or_single, Sequence) else seq_or_single
single: SomeClass = seq_or_single if isinstance(seq_or_single, SomeClass) else seq_or_single[0]
# error[invalid-assignment]: Object of type `object` is not assignable to `SomeClass`
```

While if you write the same logic in a normal if condition, you'll have the correct type inferred.

Playground: https://play.ty.dev/fb6855f8-b9ff-4c7b-bf4c-89cd79aa861a

### Version

ty 0.0.39 (020512517 2026-05-22)

**Comment by charliermarsh** [comment]
I think these are correct, because you could have a subclass of both:

```python
from collections.abc import Sequence
from typing import overload


class SomeClass:
    attribute: int = 1


class Both(SomeClass, Sequence[object]):
    @overload
    def __getitem__(self, index: int) -> object: ...
    @overload
    def __getitem__(self, index: slice) -> Sequence[object]: ...

    def __getitem__(self, index: int | slice) -> object | Sequence[object]:
        if isinstance(index, slice):
            return ()
        return "not a SomeClass"

    def __len__(self) -> int:
        return 1


seq_or_single: Sequence[SomeClass] | SomeClass = Both()

single = (
    seq_or_single[0]
    if isinstance(seq_or_single, Sequence)
    else seq_or_single
)

# At runtime, `single` is the string "not a SomeClass", not a `SomeClass`.
print(single)
print(single.attribute)  # AttributeError
```

If you mark `SomeClass` with `@final`, though, the diagnostics disappear.

**Comment by HosseyNJF** [comment]
I see, but then I'm curious that why does it work if you use a normal if-condition (not an expression)? The end result is the same.

Another thing, in the playground I sent, it fails even if you check if it inherits from `SomeClass` and not `Sequence`, even though in that case you're sure that the type is definitely SomeClass. It could be also Sequence, but you know you can access the attributes.

**Comment by charliermarsh** [comment]
I think your second test has a typo -- it should be:
```python
single_2: SomeClass = seq_or_single if isinstance(seq_or_single, SomeClass) else seq_or_single[0]
a: int = single_2.attribute
```

Which is allowed.

From there I think it's consistent? This one:

```python
single = seq_or_single[0] if isinstance(seq_or_single, Sequence) else seq_or_single
```

Is unsound because a subclass could be both `SomeClass` and `Sequence[object]`, and so `__getitem__` would return `object`, not `SomeClass`.

**Comment by HosseyNJF** [comment]
Yeah my bad 😩 Makes more sense now. Thanks for your time!
